pmd-cli/src/main/java/net/sourceforge/pmd/cli/commands/internal/PmdCommand.java

```diff
@@ -323,6 +323,16 @@ protected ExecutionResult execute() {
                 }
 
                 pmdReporter.log(Level.DEBUG, "Current classpath:\n{0}", System.getProperty("java.class.path"));
+                
+                if (showProgressBar) {
+                    if (reportFile == null) {
+                        pmdReporter.log(Level.WARN, "Progressbar rendering conflicts with reporting to STDOUT. "
+                                + "No progressbar will be shown. Try running with '-r'");
```

Comment:
For the time being I'm sticking with this behavior. In non-interactive terminals progress and errors get interleaved, making it hard to read.

For interactive terminals, the too _could_ coexist, but would require a much finer control over the order of the listeners, so we could add a listener to be run at the very first (to clear the progressbar before a renderer attempts to write a violation), and another one at the very last to redraw the updated progressbar. This way, the progressbar would always be at the bottom, until execution stops, when it will be deleted completely.

Reply:
Would it be possible just to compose listeners to define their relative order?

Eg make `Renderer#newListener()` be `Renderer#newListener(OutputMethod out)` where an `OutputMethod` is an abstraction over the actual `OutputStream` that the renderer renders to. It would have methods like `beforeOutput`, `output`, `afterOutput`. In an interactive terminal the beforeOutput removes the progress bar, allowing the renderer to append text with `output`. Then, `afterOutput` would reprint the progress bar.

Although for the time being, I think the behavior you implemented is practical. Maybe I would just clarify the message: `Try running with arguments '-r <file>' to output the report to a file.`

Reply:
But that would require surfacing the `OutputStream` / `OutputMethod` all the way to `PmdAnalysis` to let callers mingle with it… I feel it's exposing too much for something too specific…

I do take the feedback on the message however.